docs/developer-docs/security/rust-canister-development-security-best-practices.md

```diff
@@ -262,8 +262,166 @@ Every inter-canister call is guaranteed to receive a response, either from the c
+- Currently, Motoko does not allow setting custom cleanup callbacks. Cleanup is used internally by Motoko to do some cleanups.
+- As discussed above, the Rust CDK has a feature that automatically releases local variables in cleanup, which [can be used to release locks](https://internetcomputer.org/docs/current/developer-docs/security/rust-canister-development-security-best-practices#be-aware-that-there-is-no-reliable-message-ordering). Since only one cleanup callback can be defined, any custom cleanup would currently have to implement that feature itself if needed, making this currently hard to use and understand.
+
+We proceed to discuss “journaling”, which is currently our recommended way of addressing the problem at hand. 
+
+#### Journaling
+
+Journaling can be used for ensuring that tasks are completed correctly in an asynchronous context, where any instruction or async task can fail. Journaling is generally useful in any security critical application canister on the IC. The journaling concept we describe here is inspired and adapted from journaling in file systems.
+
+Conceptually, a journal is a chronological list of records kept in a canister’s storage. It keeps track of tasks before they begin and when they are completed. Before each failable task, the journal records the intent to execute the task, and after the task, the journal records the result. The journal supports idempotent task flows by providing the necessary information for the canister to resume flows that failed to complete, report progress for ongoing flows, and report results for completed flows. Retries can be initiated by calls or automatically on heartbeat. If the task flow was completed in a heartbeat, a user can take advantage of idempotency to check the result.
+
+Creating a record in the journal is called “journaling”.  For example, to make an unreliable async call to a ledger:
+
+1. Check the journal to ensure the transfer is not already in progress. If it is already in progress, go into recovery (see [Recovery](#recovery) section below). Otherwise, journal the intent to call a ledger to transfer 1 token from A to B. The journaled intent should contain sufficient context to later identify what happened to the call.
+
+   - An “in progress” transfer would show in the journal as an entry containing intent to do the transfer without an entry containing the result of the transfer call.
+
+1. Call ledger to transfer 1 token from A to B.
+
+1. Journal the result of the transfer.
+
+   - On failure, record the error.
+
+   - On success, record success. (In order to commit the record, intercanister call can be made to an endpoint on the same canister that does nothing. Otherwise a trap could erase the journaled result, complicating recovery.)
```

Comment:
"an inter-canister call"?
More importantly, should we recommend this call to be made? As-is, this sounds optional, but as discussed previously, without it a trap would render the journal useless for automated recovery (it would still be useful for manual debugging).

Reply:
In particular, the second paragraph suggests in this section suggests that the journal can be automatically replayed. Let's make it more nuanced?

Reply:
ICDex had a manual resolution mechanism that allowed some admin to retry failed transactions. I believe that is in line with our philosophy here.

> (In order to commit the record, intercanister call can be made to an endpoint on the same canister that does nothing. Otherwise a trap could erase the journaled result, complicating recovery.)

This sounds like we would benefit from a way to commit state or split messages using a system call to avoid having to use this hack? Is this a feature we want to add to the IC?

Reply:
> "an inter-canister call"? More importantly, should we recommend this call to be made? As-is, this sounds optional, but as discussed previously, without it a trap would likely render the journal useless for automated recovery (it would still be useful for manual debugging).

The journal is not useless even if a trap deletes the result. On resumption, the next task would be to obtain the result that was deleted. If it is cheap to obtain the result again and fill in the lost part of the journal, it could be beneficial to skip the extra call.

Reply:
> This sounds like we would benefit from a way to commit state or split messages using a system call to avoid having to use this hack? Is this a feature we want to add to the IC?

@roelstorms this has been raised by various people, including @oggy-dfin. I'm not aware of any plan to add it. Perhaps @oggy-dfin would be the person best positioned to push for it.

Reply:
> The journal is not useless even if a trap deletes the result. On resumption, the next task would be to obtain the result that was deleted. If it is cheap to obtain the result again and fill in the lost part of the journal, it could be beneficial to skip the extra call.
You're right, it makes sense.

> Perhaps @oggy-dfin would be the person best positioned to push for it.
I will bring it up with the runtime/execution teams but I suspect it won't make it high up the priority list unless we can get some community requests for that.